// ngx-xml2json/fesm5/ngx-xml2json.js
import { Injectable, Component, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgxXml2jsonService = /** @class */ (function () {
    function NgxXml2jsonService() {
    }
    /**
     * @param {?} xml
     * @return {?}
     */
    NgxXml2jsonService.prototype.xmlToJson = /**
     * @param {?} xml
     * @return {?}
     */
    function (xml) {
        var /** @type {?} */ obj = {};
        if (xml.nodeType === 1) {
            if (xml.attributes.length > 0) {
                obj['@attributes'] = {};
                for (var /** @type {?} */ j = 0; j < xml.attributes.length; j += 1) {
                    var /** @type {?} */ attribute = xml.attributes.item(j);
                    obj['@attributes'][attribute.nodeName] = attribute.nodeValue;
                }
            }
        }
        else if (xml.nodeType === 3) {
            obj = xml.nodeValue;
        }
        if (xml.hasChildNodes() && xml.childNodes.length === 1 && xml.childNodes[0].nodeType === 3) {
            obj = xml.childNodes[0].nodeValue;
        }
        else if (xml.hasChildNodes()) {
            for (var /** @type {?} */ i = 0; i < xml.childNodes.length; i += 1) {
                var /** @type {?} */ item = xml.childNodes.item(i);
                var /** @type {?} */ nodeName = item.nodeName;
                if (typeof (obj[nodeName]) === 'undefined') {
                    obj[nodeName] = this.xmlToJson(item);
                }
                else {
                    if (typeof (obj[nodeName].push) === 'undefined') {
                        var /** @type {?} */ old = obj[nodeName];
                        obj[nodeName] = [];
                        obj[nodeName].push(old);
                    }
                    obj[nodeName].push(this.xmlToJson(item));
                }
            }
        }
        return obj;
    };
    NgxXml2jsonService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] },
    ];
    /** @nocollapse */
    NgxXml2jsonService.ctorParameters = function () { return []; };
    /** @nocollapse */ NgxXml2jsonService.ngInjectableDef = defineInjectable({ factory: function NgxXml2jsonService_Factory() { return new NgxXml2jsonService(); }, token: NgxXml2jsonService, providedIn: "root" });
    return NgxXml2jsonService;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgxXml2jsonComponent = /** @class */ (function () {
    function NgxXml2jsonComponent() {
    }
    /**
     * @return {?}
     */
    NgxXml2jsonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    NgxXml2jsonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-ngx-xml2json',
                    template: "\n    <p>\n      Ngx-xml2json\n    </p>\n  ",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    NgxXml2jsonComponent.ctorParameters = function () { return []; };
    return NgxXml2jsonComponent;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var NgxXml2jsonModule = /** @class */ (function () {
    function NgxXml2jsonModule() {
    }
    NgxXml2jsonModule.decorators = [
        { type: NgModule, args: [{
                    imports: [],
                    declarations: [NgxXml2jsonComponent],
                    exports: [NgxXml2jsonComponent]
                },] },
    ];
    return NgxXml2jsonModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { NgxXml2jsonService, NgxXml2jsonComponent, NgxXml2jsonModule };
