// ngx-xml2json/esm5/lib/ngx-xml2json.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component } from '@angular/core';
var NgxXml2jsonComponent = /** @class */ (function () {
    function NgxXml2jsonComponent() {
    }
    /**
     * @return {?}
     */
    NgxXml2jsonComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    NgxXml2jsonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-ngx-xml2json',
                    template: "\n    <p>\n      Ngx-xml2json\n    </p>\n  ",
                    styles: []
                },] },
    ];
    /** @nocollapse */
    NgxXml2jsonComponent.ctorParameters = function () { return []; };
    return NgxXml2jsonComponent;
}());
export { NgxXml2jsonComponent };
function NgxXml2jsonComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    NgxXml2jsonComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    NgxXml2jsonComponent.ctorParameters;
}
