// ngx-xml2json/fesm2015/ngx-xml2json.js
import { Injectable, Component, NgModule, defineInjectable } from '@angular/core';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxXml2jsonService {
    constructor() { }
    /**
     * @param {?} xml
     * @return {?}
     */
    xmlToJson(xml) {
        let /** @type {?} */ obj = {};
        if (xml.nodeType === 1) {
            if (xml.attributes.length > 0) {
                obj['@attributes'] = {};
                for (let /** @type {?} */ j = 0; j < xml.attributes.length; j += 1) {
                    const /** @type {?} */ attribute = xml.attributes.item(j);
                    obj['@attributes'][attribute.nodeName] = attribute.nodeValue;
                }
            }
        }
        else if (xml.nodeType === 3) {
            obj = xml.nodeValue;
        }
        if (xml.hasChildNodes() && xml.childNodes.length === 1 && xml.childNodes[0].nodeType === 3) {
            obj = xml.childNodes[0].nodeValue;
        }
        else if (xml.hasChildNodes()) {
            for (let /** @type {?} */ i = 0; i < xml.childNodes.length; i += 1) {
                const /** @type {?} */ item = xml.childNodes.item(i);
                const /** @type {?} */ nodeName = item.nodeName;
                if (typeof (obj[nodeName]) === 'undefined') {
                    obj[nodeName] = this.xmlToJson(item);
                }
                else {
                    if (typeof (obj[nodeName].push) === 'undefined') {
                        const /** @type {?} */ old = obj[nodeName];
                        obj[nodeName] = [];
                        obj[nodeName].push(old);
                    }
                    obj[nodeName].push(this.xmlToJson(item));
                }
            }
        }
        return obj;
    }
}
NgxXml2jsonService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] },
];
/** @nocollapse */
NgxXml2jsonService.ctorParameters = () => [];
/** @nocollapse */ NgxXml2jsonService.ngInjectableDef = defineInjectable({ factory: function NgxXml2jsonService_Factory() { return new NgxXml2jsonService(); }, token: NgxXml2jsonService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxXml2jsonComponent {
    constructor() { }
    /**
     * @return {?}
     */
    ngOnInit() { }
}
NgxXml2jsonComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-ngx-xml2json',
                template: `
    <p>
      Ngx-xml2json
    </p>
  `,
                styles: []
            },] },
];
/** @nocollapse */
NgxXml2jsonComponent.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class NgxXml2jsonModule {
}
NgxXml2jsonModule.decorators = [
    { type: NgModule, args: [{
                imports: [],
                declarations: [NgxXml2jsonComponent],
                exports: [NgxXml2jsonComponent]
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

export { NgxXml2jsonService, NgxXml2jsonComponent, NgxXml2jsonModule };
